Add a game speed dropdown to the lobby options.

pchote opened the pull request:
All the recent perf improvements have made it (hopefully) feasible to implement a game speed modifier by adjusting the timestep and order latency values.  This will need testing in real network games (not just toy tests) before we consider merging it.

pchote committed:
Add Timestep property to GlobalSettings.

pchote committed:
Add game speed dropdown to the lobby.

pchote committed:
Update UI timers for variable game speed.

Mailaender commented on OpenRA.Mods.Common/UtilityCommands/ActorStatsExport.cs:
```diff
@@ -79,7 +80,9 @@ public static DataTable GenerateTable()
-							var damagePerSecond = (1000f / Game.Timestep) * (damage * burst) / (delay + burst * rateOfFire);
+
+							// Assumes the default game speed (25 ticks / sec)
+							var damagePerSecond = (1000f / 40) * (damage * burst) / (delay + burst * rateOfFire);
```
I suggest you don't leave a magic variable here as you just made it configurable and instead try:

```csharp
var gameSpeed = Game.ModData.Manifest.Get<GameSpeeds>().First();
```

pchote replied:
This would report DPS values for the slowest game speed, which isn't what users expect.

Mailaender replied:
A `bool GameSpeed.IsDefault` might help you with finding the right one.

pchote replied:
I'm not prepared to add core engine code just to work around a utility rule that its not clear whether anyone uses.

Mailaender replied:
It is certainly cleaner for the lobby code as well. Currently you are enforcing a hidden string comparison. If someone does not define a game speed named "Default", but maybe call it "Normal" like in the originals, it will also break.

pchote replied:
The names are separated from the ids to support future translation, so that isn't true.

Mailaender replied:
If you don't want to change your system and add a boolean than just let it pick the one with the "default" ID.

penev92 commented:
Played some test games with @Mailaender and noticed that the support power timers get scaled with the speed (for example Chronosphere takes 1 minute to charge) but the max times don't. So the tooltip says `01:00 / 04:00` when it starts to charge.

Mailaender commented:
Otherwise it was playable. It is more demanding, but a 2v2 comp stomp is still possible performance wise.

Mailaender commented:
Also replays work fine. :white_check_mark:

Mailaender commented on OpenRA.Game/Widgets/WidgetUtils.cs:
```diff
@@ -158,14 +158,14 @@ public static void DrawPanelPartial(string collection, Rectangle bounds, PanelSi
 		}
 
-		public static string FormatTime(int ticks)
+		public static string FormatTime(int ticks, int timestep = 40)
```
Better not allow those overloads with magic values and instead fix all occurrences.

pchote commented:
Updated.

Mailaender commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -187,6 +188,7 @@ public class Global
 			public int StartingCash = 5000;
 			public string TechLevel = "none";
 			public string StartingUnitsClass = "none";
+			public string GameSpeedType = "default";
```
What I mean is that this default type is hard-coded to a specific undocumented ID.

pchote commented:
Updated.  I changed the actor stats export to report damage per tick to remove the arbitrary ticks / second variable.  In the process I discovered that the stats export command doesn't even work, so I have also filed #9272.

Mailaender commented:
Needs a rebase.

penev92 commented:
Can't find anything obviously wrong with it, so :+1:

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/1eef037432d479349852bc7769c5ced7fe3ca862...f7402b30a11a0d52ce44f9643b73786a6c03397d)

LavenderMoon commented:
In the future, could this be set to allow changing game speed while in-game for singleplayer and skirmish games (and maybe co-op, where it could use the slowest speed of all players like in Introversion's *DEFCON*)?
